src/main/kotlin/io/sip3/salto/ce/sip/SipTransactionHandler.kt

```diff
@@ -0,0 +1,199 @@
+            recordCallUsersAttributes = it
+        }
+        config().getJsonObject("vertx")?.getInteger("instances")?.let {
+            instances = it
+        }
+
+        vertx.setPeriodic(expirationDelay) {
+            terminateExpiredTransactions()
+        }
+
+        val index = config().getInteger("index")
+        vertx.eventBus().localConsumer<Pair<Packet, SIPMessage>>(PREFIX + "_$index") { event ->
+            try {
+                val (packet, message) = event.body()
+                handle(packet, message)
+            } catch (e: Exception) {
+                logger.error("SipTransactionHandler 'handle()' failed.", e)
+            }
+        }
+    }
+
+    open fun handle(packet: Packet, message: SIPMessage) {
+        when (message.cseqMethod()) {
+            // To simplify call aggregation we decided to skip ACK and CANCEL transaction.
+            // Moreover, skipped ACK and CANCEL transactions will not affect final result.
+            "ACK", "CANCEL" -> return
+            else -> transactions.getOrPut(message.transactionId()) { SipTransaction() }.addMessage(packet, message)
+        }
+    }
+
+    open fun terminateExpiredTransactions() {
+        val now = System.currentTimeMillis()
+
+        transactions.filterValues { transaction ->
+            (transaction.terminatedAt?.let { it + terminationTimeout } ?: transaction.createdAt + aggregationTimeout) < now
+        }.forEach { (tid, transaction) ->
+            transactions.remove(tid)
+            routeTransaction(transaction)
+        }
+    }
+
+    open fun routeTransaction(transaction: SipTransaction) {
+        val prefix = when (transaction.cseqMethod) {
+            "REGISTER", "NOTIFY", "MESSAGE", "OPTIONS", "SUBSCRIBE" -> RoutesCE.sip + "_${transaction.cseqMethod.toLowerCase()}"
+            else -> RoutesCE.sip + "_call"
+        }
+
+        when (prefix) {
+            RoutesCE.sip + "_call" -> {
+                val index = transaction.callId.hashCode()
+                val route = prefix + "_${abs(index % instances)}"
+                vertx.eventBus().send(route, transaction, USE_LOCAL_CODEC)
+            }
+            RoutesCE.sip + "_register" -> {
+                // RFC-3261 10.2: The To header field contains the address of record
+                // whose registration is to be created, queried, or modified.
+                val index = (transaction.request?.toUri() ?: transaction.response?.toUri()).hashCode()
+                val route = prefix + "_${abs(index % instances)}"
+                vertx.eventBus().send(route, transaction, USE_LOCAL_CODEC)
+            }
+            RoutesCE.sip + "_message",
+            RoutesCE.sip + "_options" -> {
+                val index = transaction.callId.hashCode()
+                val route = prefix + "_${abs(index % instances)}"
+                vertx.eventBus().send(route, transaction, USE_LOCAL_CODEC)
+            }
+
+            else -> {
+                writeAttributes(transaction)
```

Comment:
I don't see much point now in separate `SipMessageHadndler` and `SipOptionsHandler` if we keep this else logic... Either we will have a separate handler always or we just have a default logic for all this transaction-like SIP messages.

Reply:
So, we agreed to remove these 2 handlers for now and leave default behavior